pkg/networkservice/common/checkconn/client_test.go

```diff
@@ -0,0 +1,143 @@
+	"go.uber.org/goleak"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/checkconn"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/monitor"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/updatepath"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/updatetoken"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/eventchannel"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/sandbox"
+)
+
+func TestCheckConn(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	eventCh := make(chan *networkservice.ConnectionEvent, 1)
+	defer close(eventCh)
+
+	monitorServer := eventchannel.NewMonitorServer(eventCh)
+
+	counter := newCounterServer()
+	server := chain.NewNetworkServiceServer(
+		updatetoken.NewServer(sandbox.GenerateTestToken),
+		updatepath.NewServer("testServer"),
+		monitor.NewServer(ctx, &monitorServer),
+		updatetoken.NewServer(sandbox.GenerateTestToken),
+		counter,
+	)
+
+	// Abort returning back connection
+	errServer := newErrorServer()
+	client := chain.NewNetworkServiceClient(
+		updatepath.NewClient("testClient"),
+		checkconn.NewClient(adapters.NewMonitorServerToClient(monitorServer)),
+		adapters.NewServerToClient(
+			chain.NewNetworkServiceServer(
+				errServer,
+				server,
+			),
+		),
+	)
+
+	// Server increments the counter but returns false on request
+	// checkconn chain element should call close and release resources
+	_, err := client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			NetworkService: "ns-1",
+		},
+	})
+	require.Error(t, err)
+	require.Equal(t, int32(0), counter.activeConnections)
+
+	// Allow requests
+	errServer.disable()
+	conn, err := client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			NetworkService: "ns-1",
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, int32(1), counter.activeConnections)
+
+	_, err = client.Close(ctx, conn)
+	require.NoError(t, err)
+	require.Equal(t, int32(0), counter.activeConnections)
+}
+
+// errorServer - simulates error on the server side
+type errorServer struct {
+	enabled bool
+}
+
+func newErrorServer() *errorServer {
+	return &errorServer{enabled: true}
+}
+
+func (s *errorServer) disable() {
+	s.enabled = false
+}
+
+func (s *errorServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	conn, e := next.Server(ctx).Request(ctx, request.Clone())
+	if s.enabled {
+		return nil, errors.New("error on the server")
```

Comment:
Can you please also cancel here Request context, to make sure that Close is not based on the same canceled context?